packages/react-aria-components/test/Tabs.test.js

```diff
@@ -474,4 +474,92 @@ describe('Tabs', () => {
     expect(innerTabs[0]).toHaveTextContent('One');
     expect(innerTabs[1]).toHaveTextContent('Two');
   });
+
+  it('can add tabs and keep the current selected key', async () => {
+    let onSelectionChange = jest.fn();
+    function Example(props) {
+      let [tabs, setTabs] = useState([
+        {id: 1, title: 'Tab 1', content: 'Tab body 1'},
+        {id: 2, title: 'Tab 2', content: 'Tab body 2'},
+        {id: 3, title: 'Tab 3', content: 'Tab body 3'}
+      ]);
+
+      const [selectedTabId, setSelectedTabId] = useState(tabs[0].id);
+
+      let addTab = () => {
+        const tabId = tabs.length + 1;
+
+        setTabs((prevTabs) => [
+          ...prevTabs,
+          {
+            id: tabId,
+            title: `Tab ${tabId}`,
+            content: `Tab body ${tabId}`
+          }
+        ]);
+
+        // Use functional update to ensure you're working with the most recent state
+        setSelectedTabId(tabId);
+      };
+
+      let removeTab = () => {
+        if (tabs.length > 1) {
+          setTabs((prevTabs) => {
+            const updatedTabs = prevTabs.slice(0, -1);
+            // Update selectedTabId to the last remaining tab's ID if the current selected tab is removed
+            const newSelectedTabId = updatedTabs[updatedTabs.length - 1].id;
+            setSelectedTabId(newSelectedTabId);
+            return updatedTabs;
+          });
+        }
+      };
+
+      const onSelectionChange = (value) => {
+        setSelectedTabId(value);
+        props.onSelectionChange(value);
+      };
+
+      return (
+        <Tabs selectedKey={selectedTabId} onSelectionChange={onSelectionChange}>
+          <div style={{display: 'flex'}}>
+            <TabList aria-label="Dynamic tabs" items={tabs} style={{flex: 1}}>
+              {(item) => (
+                <Tab>
+                  {({isSelected}) => (
+                    <p
+                      style={{
+                        color: isSelected ? 'red' : 'black'
+                      }}>
+                      {item.title}
+                    </p>
+                  )}
+                </Tab>
+              )}
+            </TabList>
+            <div className="button-group">
+              <Button onPress={addTab}>Add tab</Button>
+              <Button onPress={removeTab}>Remove tab</Button>
+            </div>
+          </div>
+          <Collection items={tabs}>
+            {(item) => (
+              <TabPanel
+                style={{
+                  borderTop: '2px solid black'
+                }}>
+                {item.content}
+              </TabPanel>
+            )}
+          </Collection>
+        </Tabs>
+      );
+    }
+    render(<Example onSelectionChange={onSelectionChange} />);
+    await user.tab();
+    await user.keyboard('{ArrowRight}');
+    await user.tab();
+    onSelectionChange.mockClear();
+    await user.keyboard('{Enter}');
+    expect(onSelectionChange).not.toHaveBeenCalled();
```

Comment:
Just making sure I'm following the test,
- Three tabs are rendered, focus on "body".
- User presses "tab" to focus the tabs, first tab selected out of three tabs.
- User moves from the first tab to the second tab.
- User presses "tab" and we're on the "Add button".
- Users pressed the "Add button" and a tab is added. Selection remains on the second tab of four tabs.

Why did you create a remove function if it isn't used?

Reply:
original component used in test is just a copy from the reproduction codesandbox, but I've taken it as an opportunity now to add a few more assertions